src/Cloud/weglide/UploadIGCFile.cpp

```diff
@@ -0,0 +1,241 @@
+  try {
+    json = co_await UploadFlight(*Net::curl, uri, pilot, glider_id,
+      igc_path, progress);
+  }
+  catch (const std::exception& e) {
+    // Passing the exception without 'Co::InvokeTask' ???
+    throw std::runtime_error(e.what());
+  }
+}
+#else
+static Co::InvokeTask
+UpdateTask(Path igc_path, const char* uri, const struct Pilot& pilot,
+           uint_least32_t glider_id, boost::json::value& json,
+           ProgressListener& progress) {
+  json = co_await UploadFlight(*Net::curl, uri, pilot, glider_id,
+    igc_path, progress);
+}
+#endif
+
+static void
+UploadJsonInterpreter(const boost::json::value &json,
+                      struct UploadResponse& response) {
+  if (json.is_array() && (json.get_array().size() > 0)) {
+    auto flight_obj = json.get_array().at(0);
+    response.scoring_date = GetJsonString(flight_obj, "scoring_date").c_str();
+    response.flight_id = GetJsonInt(flight_obj, "id");
+    response.registration = GetJsonString(flight_obj, "registration").c_str();
+    response.competition_id = GetJsonString(flight_obj, "competition_id").c_str();
+    auto user = flight_obj.at("user");
+    if (user.is_object()) {
+      auto user_obj = user.as_object();
+      response.pilot.id = GetJsonInt(user_obj, "id");
+      response.pilot.name = GetJsonString(user_obj, "name").c_str();
+    }
+    auto aircraft = flight_obj.at("aircraft");
+    if (aircraft.is_object()) {
+      auto aircraft_obj = aircraft.as_object();
+      response.glider.id = GetJsonInt(aircraft_obj, "id");
+      response.glider.name = GetJsonString(aircraft_obj, "name").c_str();
+      response.glider.kind = GetJsonString(aircraft_obj, "kind").c_str();
+      response.glider.sc_class = GetJsonString(aircraft_obj, "sc_class").c_str();
+    }
+  }
+  else {
+    throw std::runtime_error("The http response has a wrong JSON format!");
+  }
+}
+
+static void
+UploadSuccessDialog(const struct UploadResponse& response, const Path igc_path) {
+  // TODO: Create a real Dialog with fields
+  // With this Dialog insert the possibilty to update/patch the flight
+  // f.e. copilot in double seater, scoring class, short comment and so on
+  StaticString<0x1000> msg;
+  msg.Format(_("File upload '%s' was succesfull"),
+    igc_path.c_str());
+  msg.append(_T("!\n\n"));
+  msg.AppendFormat(_T("%-20s: %u\n"), _("flight_id"), response.flight_id);
+  msg.AppendFormat(_T("%-20s: %s\n"), _("scoring_date"),
+    response.scoring_date.c_str());
+  msg.AppendFormat(_T("%-20s: %s (%d)\n"), _("pilot"),
+    response.pilot.name.c_str(), response.pilot.id);
+  msg.AppendFormat(_T("%-20s: %s (%u)\n"), _("aircraft type"),
+    response.glider.name.c_str(), response.glider.id);
+  msg.AppendFormat(_T("%-20s: %s: %s "), _("glider"), _("reg"),
+    response.registration.c_str());
+  msg.AppendFormat(_T("%s: %s\n"), _("cid"),
+    response.competition_id.c_str());
+
+  ShowMessageBox(msg, _T("WeGlide Upload"), MB_OK | MB_ICONEXCLAMATION);
```

Comment:
So... is this a backend protocol library, or is this a frontend UI library? This looks like you're mixing the two again.

Reply:
Because unfortunately I don't understand your frontend and backend area  up to now. For me the content of the Cloud folder belongs to the frontend, but so I will split this functionality again to fullfill your requirements.
So I will add a `src/contest `folder to placing the frontend stuff here.
Maybe this folder shoud be a part of `src/cloudserver/contest` - and `src/Tracking` should be moved to `src/cloudserver/tracking` too because this code belongs to the same root - but this is a later story...

Reply:
You don't understand the distiction because your code is a mix of all layers, without any modularity. That way, the WeGlide client can't be reused for other parts of XCSoar. For example, you can't debug it on the command line.

Reply:
Yes, I know ;-( Hopefully now I think to understand what you mean. My question is: Can I do that in one folder or it is recommended to split this?